### PR TITLE
Fix additional constants embedded in instruction stream on AMD64.

### PR DESCRIPTION
**Pull request by lateralusX**
A couple of more op codes embedded constant values in instruction stream. When using MONO_ARCH_CODE_EXEC_ONLY these needs to be moved into GOT slots.

Thanks @buggeststar for bring the issues to our attention.

### CODE REVIEW

**Comment by lateralusX** [comment]
Yes, I looked through mini-amd64.c for constructs using constants reading using a relative addressing using AMD64_RIP.

For the complete set of current fixes its code_exec_only for JIT and MONO_ARCH_CODE_EXEC_ONLY for AOT compiler related changes.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
The max instruction lengths might need to be updated.

File: mono/mini/mini-amd64.c [hunk 1]
```diff
@@ -6048,19 +6048,33 @@ mono_arch_output_basic_block (MonoCompile *cfg, MonoBasicBlock *bb)
 			static double r8_0 = -0.0;
 
 			g_assert (ins->sreg1 == ins->dreg);
-					
-			mono_add_patch_info (cfg, offset, MONO_PATCH_INFO_R8, &r8_0);
-			amd64_sse_xorpd_reg_membase (code, ins->dreg, AMD64_RIP, 0);
+
+			if (cfg->compile_aot && cfg->code_exec_only) {
+				mono_add_patch_info (cfg, offset, MONO_PATCH_INFO_R8_GOT, &r8_0);
+				amd64_mov_reg_membase (code, AMD64_R11, AMD64_RIP, 0, sizeof(gpointer));
```
Comment (jaykrell) [inline mono/mini/mini-amd64.c]: Space after `sizeof`, but really, host/target confusion.

File: mono/mini/mini-amd64.c [hunk 2]
```diff
@@ -6048,19 +6048,33 @@ mono_arch_output_basic_block (MonoCompile *cfg, MonoBasicBlock *bb)
...
 
 			g_assert (ins->sreg1 == ins->dreg);
-					
-			mono_add_patch_info (cfg, offset, MONO_PATCH_INFO_R8, &d);
-			amd64_sse_andpd_reg_membase (code, ins->dreg, AMD64_RIP, 0);
-			break;		
+
+			if (cfg->compile_aot && cfg->code_exec_only) {
+				mono_add_patch_info (cfg, offset, MONO_PATCH_INFO_R8_GOT, &d);
+				amd64_mov_reg_membase (code, AMD64_R11, AMD64_RIP, 0, sizeof(gpointer));
```
Comment (jaykrell) [inline mono/mini/mini-amd64.c]: Space after sizeof, but really, host/target confusion.

File: mono/mini/mini-amd64.c [hunk 3]
```diff
@@ -6083,8 +6097,15 @@ mono_arch_output_basic_block (MonoCompile *cfg, MonoBasicBlock *bb)
 
 			g_assert (ins->sreg1 == ins->dreg);
 
-			mono_add_patch_info (cfg, offset, MONO_PATCH_INFO_R4, &r4_0);
-			amd64_sse_movss_reg_membase (code, MONO_ARCH_FP_SCRATCH_REG, AMD64_RIP, 0);
+			if (cfg->compile_aot && cfg->code_exec_only) {
+				mono_add_patch_info (cfg, offset, MONO_PATCH_INFO_R4_GOT, &r4_0);
+				amd64_mov_reg_membase (code, AMD64_R11, AMD64_RIP, 0, sizeof(gpointer));
```
Comment (jaykrell) [inline mono/mini/mini-amd64.c]: Space after sizeof, but really, host/target confusion.

Comment (jaykrell) [reply mono/mini/mini-amd64.c]: Can the check of `code_exec_only` be removed, and just always do this, at least for AOT?

Comment (lateralusX) [reply mono/mini/mini-amd64.c]: It was decided to keep this feature as an opt in feature at the moment.

**Comment by lateralusX** [comment]
> The max instruction lengths might need to be updated.

@vargaz Looked through the affected instructions and they all seems to be large enough.

**Comment by lateralusX** [comment]
@monojenkins rebuild failed

**Pull request merged**